### PR TITLE
feat(edit-column): initial implementation based on prototype

### LINKED ISSUE #3: Implement the vaadin-grid-edit-column component
- Extend the vaadin-grid-column element
- Should have a default renderer that makes the cells editable
- Should fire an event ("item-property-changed") when the editor closes
  - Update the in-memory item synchronously (app logic is responsible of reverting the change)

DoD: https://docs.google.com/spreadsheets/d/1F6KLLJDKbw3EhdmIzloPNyDKFOGFZieoDR5eMywAioE/edit#gid=1117754529
Proto: https://www.dropbox.com/s/jq79ew1vbg8rpjt/inline-grid.html?dl=0

screenshot 2018-12-12 09 35 05

### PR DESCRIPTION
**Pull request by web-padawan**
Fixes #3 

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid-pro/8)

### CODE REVIEW

**Comment by web-padawan** [comment]
Dismissed the second comment from the CLA assistant, as the CLA actually signed.

File: src/vaadin-pro-grid-inline-editing-mixin.html [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      if (shouldResize) {
+        this.notifyResize();
+      }
+    }
+
+    _switchEditCell(e) {
+      this._cancelStopEdit();
+
+      const columnTreeLevel = this._columnTree.length - 1;
+      const cols = this._columnTree[columnTreeLevel].filter(this._isEditColumn);
```
Comment (tomivirkki) [inline src/vaadin-pro-grid-inline-editing-mixin.html]: I think you'll need to sort these by column `_order` which is the current effective order (drag-reordering doesn't actually reorder the column elements in the DOM). Also, some columns might be hidden so those should be filtered out. [Example](https://github.com/vaadin/vaadin-grid/blob/805e1158795f3563238d49c058dbf91c29b3e4ca/src/vaadin-grid-column-resizing-mixin.html#L44)

File: src/vaadin-pro-grid-inline-editing-mixin.html [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    ready() {
+      super.ready();
+
+      this.addEventListener('item-property-changed', e => {
+        const {path, value, index, item} = e.detail;
+        this.set(path, value, item);
+        this.notifyPath('items.' + index + '.' + path, value);
+      });
+
+      this.addEventListener('click', e => {
```
[the remark this answers is not in the document]
Comment (tomivirkki) [reply src/vaadin-pro-grid-inline-editing-mixin.html]: Yes, doubleclick should be the default

File: src/vaadin-pro-grid-edit-column.html [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        };
+      }
+
+      ready() {
+        super.ready();
+
+        this.__editStateTemplate = this._prepareTemplatizer(this.$.editStateTemplate);
+        // needed to override the dataHost correctly in case internal template is used.
+        this.__editStateTemplate.templatizer.dataHost = this;
+        this.__editStateRenderer = this.constructor._getEditorRenderer();
```
Comment (tomivirkki) [inline src/vaadin-pro-grid-edit-column.html]: As you scroll the grid while the cell is in edit mode, the same editor component will appear on an arbitrary row with the content of the cell being actually edited

Comment (web-padawan) [reply src/vaadin-pro-grid-edit-column.html]: Discussed in person and it is OK for now to exit edit state when scrolled out of the viewport.

File: src/vaadin-pro-grid-edit-column.html [hunk 4]
```diff
@@ -0,0 +1,190 @@
+<!--
+@license
+Copyright (c) 2018 Vaadin Ltd.
+This program is available under Commercial Vaadin Add-On License 3.0 (CVALv3).
+See <a href="https://vaadin.com/license/cval-3">the website</a> for the complete license.
+-->
+
+<link rel="import" href="../../polymer/lib/elements/dom-if.html">
```
[the remark this answers is not in the document]
Comment (web-padawan) [reply src/vaadin-pro-grid-edit-column.html]: Removed.

File: src/vaadin-pro-grid-edit-column.html [hunk 5]
```diff
@@ -0,0 +1,164 @@
+<!--
+@license
+Copyright (c) 2018 Vaadin Ltd.
+This program is available under Commercial Vaadin Add-On License 3.0 (CVALv3).
+See <a href="https://vaadin.com/license/cval-3">the website</a> for the complete license.
+-->
+
+<link rel="import" href="../../vaadin-grid/vaadin-grid-column.html">
```
Comment (tomivirkki) [inline src/vaadin-pro-grid-edit-column.html]: Should import from src

Comment (web-padawan) [reply src/vaadin-pro-grid-edit-column.html]: Done.

File: src/vaadin-pro-grid-inline-editing-mixin.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (web-padawan) [reply src/vaadin-pro-grid-inline-editing-mixin.html]: Done.

File: src/vaadin-pro-grid-edit-column.html (hunk 3 above)
Comment (web-padawan) [reply src/vaadin-pro-grid-edit-column.html]: Done.

File: src/vaadin-pro-grid-inline-editing-mixin.html [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      if (value !== this._getItemValue(column, model)) {
+        this.dispatchEvent(new CustomEvent('item-property-changed', {
+          detail: {
+            index: model.index,
+            item: model.item,
+            path: column.path,
+            value: value
+          },
+          bubbles: true,
+          composed: true
```
[the remark this answers is not in the document]
Comment (web-padawan) [reply src/vaadin-pro-grid-inline-editing-mixin.html]: Done.

File: demo/pro-grid-basic-demos.html [hunk 7]
```diff
@@ -6,13 +6,54 @@
...
+      <template preserve-content>
+        <vaadin-pro-grid>
+          <vaadin-pro-grid-edit-column path="name.first" header="First name">
+            <template>
+              <span style="font-weight: bold;">[[item.name.first]]</span>
+            </template>
+          </vaadin-pro-grid-edit-column>
+          <vaadin-grid-column header="Last name (view)">
+            <template>
+              <span style="color: #999;">[[item.name.last]]</span>
```
[the remark this answers is not in the document]
Comment (web-padawan) [reply demo/pro-grid-basic-demos.html]: Will make more sense once we have `editorRenderer`, so let me improve it in #14

File: src/vaadin-pro-grid-inline-editing-mixin.html [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+      let nextCol = null;
+      let nextIdx = index;
+
+      // enter: move up / down
+      if (e.keyCode === 13) {
+        nextCol = column;
+        nextIdx = e.shiftKey ? index - 1 : index + 1;
+      }
+
+      // tab: move right / left
```
Comment (tomivirkki) [inline src/vaadin-pro-grid-inline-editing-mixin.html]: Seems that the edit mode is lost on tab press. Worked with a previous revision.

Comment (web-padawan) [reply src/vaadin-pro-grid-inline-editing-mixin.html]: Yep, that is by design:

> Hitting tab while editing a cell moves the focus in unedited mode to the cell to the right

It works the same way in Google Spreadsheet, I mentioned this in the morning.

**Comment by web-padawan** [comment]
Based on the discussion with @tomivirkki, @platosha and @Peppe, the keyboard navigation behavior has been slightly changed to introduce less breaking changes:

1. Added `preserveEditMode` flag (default: false) to activate edit mode on the next focused cell

2. Added `allowEnterRowChange` flag (default: false) to move focus to the next / previous row on <kbd>Enter</kbd> / <kbd>Shift</kbd> + <kbd>Enter</kbd>

Note that making <kbd>Tab</kbd> behavior in edit mode opt-in also is trickier, due to how it is currently implemented `KeyboardNavigationMixin`. Submitted #17 to handle that in a separate PR

**Comment by web-padawan** [comment]
@limonte I'm unable to reproduce Esc issue with my virtual machine (EdgeHTML 17.17134)
Looks like it only happens after upgrading to 18. Could you please submit a separate issue for that?

**Comment by limonte** [comment]
> Looks like it only happens after upgrading to 18. Could you please submit a separate issue for that?

Done, #18

**Pull request merged**